buffer/src/main/java/io/netty5/buffer/api/DefaultCompositeBuffer.java

```diff
@@ -1951,4 +1965,183 @@ public boolean process(int index, WritableComponent component) {
+        T currentComponent;
+        int pastOffset;
+        int currentReadSkip;
+        int currentWriteSkip;
+
+        private NextComponent(DefaultCompositeBuffer compositeBuffer,
+                              Function<Buffer, ComponentIterator<T>> intoIterator) {
+            this.compositeBuffer = compositeBuffer;
+            this.intoIterator = intoIterator;
+            bufs = compositeBuffer.bufs;
+            nextIndex = 1;
+            currentItr = intoIterator.apply(bufs[0]);
+            currentComponent = currentItr.first();
+        }
+
+        @SuppressWarnings("unchecked")
+        @Override
+        public <N extends Next> N next() {
+            if (currentComponent == null) {
+                return null; // Already at the end of the iteration.
+            }
+            currentComponent = currentComponent.next();
+            if (currentComponent == null) {
+                currentItr.close();
+                if (nextIndex < bufs.length) {
+                    pastOffset += bufs[nextIndex - 1].capacity();
+                    currentReadSkip = 0;
+                    currentWriteSkip = 0;
+                    currentItr = intoIterator.apply(bufs[nextIndex]);
+                    currentComponent = currentItr.first();
+                    if (currentComponent != null) {
+                        nextIndex++;
+                        return (N) this;
+                    }
+                }
+                currentItr = null;
+            }
+            return null;
+        }
+
+        @Override
+        public boolean hasReadableArray() {
+            return ((ReadableComponent) currentComponent).hasReadableArray();
+        }
+
+        @Override
+        public byte[] readableArray() {
+            return ((ReadableComponent) currentComponent).readableArray();
+        }
+
+        @Override
+        public int readableArrayOffset() {
+            return ((ReadableComponent) currentComponent).readableArrayOffset();
+        }
+
+        @Override
+        public int readableArrayLength() {
+            return ((ReadableComponent) currentComponent).readableArrayLength();
+        }
+
+        @Override
+        public long readableNativeAddress() {
+            return ((ReadableComponent) currentComponent).readableNativeAddress();
+        }
+
+        @Override
+        public ByteBuffer readableBuffer() {
+            return ((ReadableComponent) currentComponent).readableBuffer();
+        }
+
+        @Override
+        public int readableBytes() {
+            return ((ReadableComponent) currentComponent).readableBytes();
+        }
+
+        @Override
+        public ByteCursor openCursor() {
+            return ((ReadableComponent) currentComponent).openCursor();
+        }
+
+        @Override
+        public void skipReadable(int byteCount) {
+            if (byteCount < 0) {
+                throw new IllegalArgumentException("Byte count cannot be negative: " + byteCount);
+            }
```

Comment:
nit: I guess we could remove this check above as the `ReadableComponent` should check itself anyway ?